044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,295 @@
+   3. throttle if available ratio less than threshold on any volume
+4. Introduce extension points in the quotas-plugin to support pluggable sources of quotas and throttle factors
+5. Introduce a fallback throttle factor in case we cannot retrieve the cluster state
+6. strimzi-kafka-operator becomes responsible for configuring the admin client connection properties of the plugin
+
+So every broker will make its own independent throttling decision based on knowledge of the volumes on all active broker nodes.
+The brokers should all operate on a similar view of the cluster state and make a deterministic decision. If a broker detects that
+any volume in the cluster is becoming too full it will throttle production of messages. The kafka quota API isn't rich 
+enough to do anything smarter about only throttling writes to the brokers running out of space, so we fence the whole cluster.
+
+### Caveat - KRaft Disk Usage
+
+This proposal will only help prevent running out-of-disk caused by topic data. It will not prevent disks being exhausted
+due to writes to the upcoming KRaft metadata log.
+
+Currently, when using separate controller-only nodes*, those nodes are not described by `Admin#describeCluster` and we cannot use 
+`Admin#describeLogDirs` against the controllers. So the disk usage of the volume could be invisible.
+
+When using controller+broker mode*, by default the metadata log is kept in the first `log.dirs` directory but could be 
+configured to a custom location, potentially on its own volume. If it was put on a separate volume it would not be
+described in the describeLogDirs response, nor contribute to the volume usage of an existing log dir.
+
+So in some cases with controller+broker mode it would afford some protection, as growth of the metadata log dir could
+cause throttling of topic writes (because they are on the same volume as another log dir).
+
+Even if we had reliable insight into the disk usage (by extending the admin apis for example), of the volume the metadata dir 
+resides on, we cannot take effective measures from the quota plugin. Potentially you could use an Authorization
+plugin instead and block metadata-generating operations.
+
+*tested with kafka 3.3.1
+### High level changes
+To better support external sources for managing quotas this proposal introduces some new concepts to the plugin:
+
+#### Quota Source
+
+We propose adding a Quota Source concept to the plugin to provide an extension point where we could plug in
+future sources of quotas, like pull them from an external system.
+
+#### Throttle Factor
+
+We introduce the concept of throttle factor. This is a factor in the range (0.0, 1.0) that is applied to a quota to
+calculate the final byte rate limit. 
+
+This is a concept that was already implicit in the current calculations, but we want to name it so that we can use it in 
+metric names and provide an extension point in case we want to externalise the Throttle Factor.
```

Comment:
"externalise the Throttle Factor" what does that mean?

Reply:
Providing some point at which we could swap a locally calculated factor to apply for one calculated by a "leader" (or even an external service) and then consumed and applied by each broker.